pkg/operator/controller/nodeobservability/daemonset.go

```diff
@@ -208,3 +268,136 @@ func (r *NodeObservabilityReconciler) desiredDaemonSet(nodeObs *v1alpha1.NodeObs
 func labelsForNodeObservability(name string) map[string]string {
 	return map[string]string{"app": "nodeobservability", "nodeobs_cr": name}
 }
+
+// haveVolumesChanged if the current volumes differs from the desired volumes
+func haveVolumesChanged(current []corev1.Volume, desired []corev1.Volume) bool {
+	if len(current) != len(desired) {
+		return true
+	}
+	for i := 0; i < len(current); i++ {
+		cv := current[i]
+		dv := desired[i]
+		if cv.Name != dv.Name {
+			return true
+		}
+		if dv.Secret != nil {
+			if cv.Secret == nil {
+				return true
+			}
+			if cv.Secret.SecretName != dv.Secret.SecretName {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// hasContainerChanged checks if the current container differs from the
+// desired container
+func hasContainerChanged(current, desired corev1.Container) bool {
+	if current.Image != desired.Image {
+		return true
+	}
+	if !cmp.Equal(current.Args, desired.Args) {
+		return true
+	}
+
+	if hasSecurityContextChanged(current.SecurityContext, desired.SecurityContext) {
+		return true
+	}
+
+	if len(current.Env) != len(desired.Env) {
+		return true
+	}
+	currentEnvs := indexedContainerEnv(current.Env)
+	for _, e := range desired.Env {
+		if ce, ok := currentEnvs[e.Name]; !ok {
+			return true
+		} else if !cmp.Equal(ce, e) {
+			return true
+		}
+	}
+
+	if len(current.VolumeMounts) != len(desired.VolumeMounts) {
+		return true
+	}
+
+	for i := 0; i < len(current.VolumeMounts); i++ {
+		cvm := current.VolumeMounts[i]
+		dvm := desired.VolumeMounts[i]
+		if cvm.Name != dvm.Name || cvm.MountPath != dvm.MountPath {
+			return true
+		}
+	}
+
+	return false
+}
+
+func indexedContainerEnv(envs []corev1.EnvVar) map[string]corev1.EnvVar {
+	indexed := make(map[string]corev1.EnvVar)
+	for _, e := range envs {
+		indexed[e.Name] = e
+	}
+	return indexed
+}
+
+func hasSecurityContextChanged(current, desired *corev1.SecurityContext) bool {
```

Comment:
This function checks a bit more than the desired state really needs. It's more appropriate for a POD which got into the `restricted` POD security profile.

Reply:
I agree, I'm fine either ways retaining (future proof) or strip it down to whats required.

Reply:
It's tempting to keep this but I'd use [the Occam's razor principle](https://en.wikipedia.org/wiki/Occam%27s_razor) and would bring the needed things when they are needed.